### PR TITLE
#446 List journal with final marks

### LINKED ISSUE #446: List journal with final marks

### PR DESCRIPTION
**Pull request by n-gb**
Resolves #446 
screenshot from 2017-07-26 16-09-36

### CODE REVIEW

**Comment by sveredyuk** [comment]
@NadiaGubernator Don't show "0" for presence, just empty space

File: components/courses/app/finders/courses/journal/progresses_for_student.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+module Courses
+  module Journal
+    class ProgressesForStudent < ApplicationFinder
+      def initialize(student)
+        @student = student
+      end
+
+      def call
+        ::Courses::Progress.where(student_id: student.id)
```
Comment (sveredyuk) [inline components/courses/app/finders/courses/journal/progresses_for_student.rb]: Over engineering. Maybe just `student.progresses`?

File: components/courses/app/finders/courses/journal/student_mark_for_lecture.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      def call
+        mark
+      end
+
+      private
+
+      attr_reader :student, :lecture
+
+      def mark
+        if lecture.progresses.find_by(student_id: student.id)
```
Comment (sveredyuk) [inline components/courses/app/finders/courses/journal/student_mark_for_lecture.rb]: lecture.progresses.students.find(student.id)

Comment (sveredyuk) [reply components/courses/app/finders/courses/journal/student_mark_for_lecture.rb]: move to method and `||=` it

File: components/courses/app/finders/courses/journal/student_mark_for_lecture.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+          lecture.progresses.find_by(student_id: student.id).homework_mark
```
Comment (sveredyuk) [inline components/courses/app/finders/courses/journal/student_mark_for_lecture.rb]: reuse only ||= method like `student_progress.homework_mark`

File: components/courses/app/finders/courses/journal/student_presence_for_lecture.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+      def call
+        presence
+      end
+
+      private
+
+      attr_reader :student, :lecture
+
+      def presence
+        if lecture.progresses.find_by(student_id: student.id)
```
Comment (sveredyuk) [inline components/courses/app/finders/courses/journal/student_presence_for_lecture.rb]: same as previous file

File: components/courses/app/helpers/admin/courses/course_helper.rb [hunk 5]
```diff
@@ -49,6 +49,12 @@ def admin_courses_season_test_task_index_link
           class: admin_courses_active_item?('test_task')
       end
 
+      def admin_courses_season_journal_link
+        link_to icon(:'list layout', t('courses.journal.singular')),
+                admin_courses_season_journal_index_path(current_season),
```
Comment (sveredyuk) [inline components/courses/app/helpers/admin/courses/course_helper.rb]: why index_path ?

File: components/courses/app/helpers/admin/courses/journal_helper.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    module JournalHelper
+      def journal_mark(student, lecture)
+        ::Courses::Journal::StudentMarkForLecture.call(student, lecture)
+      end
+
+      def journal_presence(student, lecture)
+        ::Courses::Journal::StudentPresenceForLecture.call(student, lecture)
+      end
+
+      def add_colour(thing)
```
Comment (sveredyuk) [inline components/courses/app/helpers/admin/courses/journal_helper.rb]: courses_journal_data_color

Comment (sveredyuk) [reply components/courses/app/helpers/admin/courses/journal_helper.rb]: thing -> value

File: components/courses/app/helpers/admin/courses/journal_helper.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        elsif thing > 0
+          '#e0ffe7'
+        elsif thing == 0
+          '#fffde0'
+        elsif thing < 0
+          '#ffe8e0'
+        end
+      end
+
+      def create_id(student, lecture = nil, word = 'total')
```
Comment (sveredyuk) [inline components/courses/app/helpers/admin/courses/journal_helper.rb]: journal_value_id

File: components/courses/app/helpers/admin/courses/progress_helper.rb [hunk 8]
```diff
@@ -59,13 +59,11 @@ def set_lecture_presence(season, lecture, student)
       end
 
       def homework_mark(lecture, student)
-        return '-' unless homework_present?(lecture, student)
```
Comment (sveredyuk) [inline components/courses/app/helpers/admin/courses/progress_helper.rb:62]: that was a good idea, left view things for view helpers.
and finders/services should not do this

File: components/courses/app/helpers/admin/courses/progress_helper.rb [hunk 9]
```diff
@@ -59,13 +59,11 @@ def set_lecture_presence(season, lecture, student)
...
-
         student_progress(lecture, student).homework_mark
       end
 
       def set_mark_button(season, lecture, student, mark)
-        return '-' unless homework_present?(lecture, student)
+        return '-' unless homework_present?(lecture, student) || mark == -1
```
Comment (sveredyuk) [inline components/courses/app/helpers/admin/courses/progress_helper.rb]: magic -1 ?

File: components/courses/app/services/courses/journal/total.rb [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        total
+      end
+
+      private
+
+      attr_reader :student
+
+      def marks_total
+        sum = 0
+        progresses.each do |progress|
```
Comment (sveredyuk) [inline components/courses/app/services/courses/journal/total.rb]: `progresses.map(&: homework_mark).compact` and then no unless need... you know what to do

File: components/courses/app/services/courses/journal/total.rb [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+          sum += progress.homework_mark unless progress.homework_mark.nil?
+        end
+        sum
+      end
+
+      def presences_total
+        sum = 0
+        progresses.each do |progress|
+          sum += progress.lecture_presence unless progress.homework_mark.nil?
```
Comment (sveredyuk) [inline components/courses/app/services/courses/journal/total.rb]: :(

File: components/courses/app/services/courses/student/total_hash.rb [hunk 12]
```diff
@@ -0,0 +1,25 @@
+module Courses
+  class Student < ApplicationRecord
+    class TotalHash < ApplicationService
+      def initialize(students)
+        @students   = students
+        @total_hash = Hash.new
+      end
+
+      def call
+        students.each do |student|
```
Comment (sveredyuk) [inline components/courses/app/services/courses/student/total_hash.rb]: zstudents.each_with_object({}) do |student, total_hash|`

File: components/courses/app/views/admin/courses/journal/index.slim [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+          - lectures.each_with_index do |lecture, index|
+            th.center.aligned = 'M'
+            th.center.aligned = 'P'
+
+    tbody
+      - total_hash.each do |student, total|
+        tr
+          td = student.full_name
+          - lectures.each do |lecture|
+            td.center.aligned id=create_id(student, lecture, 'm') bgcolor = add_colour(journal_mark(student, lecture))
```
Comment (sveredyuk) [inline components/courses/app/views/admin/courses/journal/index.slim]: id = create_id(student, lecture, 'm')

File: components/courses/app/views/admin/courses/journal/index.slim [hunk 14]
```diff
@@ -0,0 +1,27 @@
+- title t('courses.journal.singular')
+
+= admin_courses_seasons_nav do
+
+  table.ui.celled.complex.table
+    thead
+      tr
+        th rowspan = "2" = t('courses.students.name')
```
Comment (sveredyuk) [inline components/courses/app/views/admin/courses/journal/index.slim]: '2'

File: components/courses/config/routes.rb [hunk 15]
```diff
@@ -24,6 +24,7 @@
           resources :progress, only:   %i[index update]
         end
```
Comment (sveredyuk) [inline components/courses/config/routes.rb:29]: enter

File: components/courses/spec/features/admin/journal/index_spec.rb [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+  let!(:first_progress)  { create(:progress, student: first_student, lecture: first_lecture, homework_mark: 1, lecture_presence: 0) }
+  let!(:second_progress) { create(:progress, student: second_student, lecture: first_lecture, homework_mark: 1, lecture_presence: -1) }
+  let!(:third_progress)  { create(:progress, student: first_student, lecture: second_lecture, homework_mark: -1, lecture_presence: -1) }
+  let!(:fourth_progress) { create(:progress, student: second_student, lecture: second_lecture, homework_mark: 0, lecture_presence: 0) }
+
+  let(:test_j_path)      { '/admin/courses/seasons/test-season/journal' }
+
+  before { visit test_j_path }
+
+  context 'displays journal correctly' do
```
Comment (sveredyuk) [inline components/courses/spec/features/admin/journal/index_spec.rb]: it

File: components/courses/spec/features/admin/journal/index_spec.rb [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+    it 'shows first student first lecture mark' do
```
Comment (sveredyuk) [inline components/courses/spec/features/admin/journal/index_spec.rb]: `# shows first student first lecture mark`

File: components/courses/app/helpers/admin/courses/course_helper.rb (hunk 5 above)
Comment (n-gb) [reply components/courses/app/helpers/admin/courses/course_helper.rb]: Because `rake routes | grep journal` gives me this route

**Review by sveredyuk — APPROVED** [review]

File: components/courses/config/routes.rb [hunk 18]
```diff
@@ -15,15 +15,18 @@
 
       resources :seasons,      except: :destroy do
         resources :mentors,    except: %i[show edit update]
+        resources :test_task,  only:   %i[index update]
+        resources :students,   only:   %i[index update]
         resources :questions,  except: :destroy
+        resources :journal,    only:   :index
```
Comment (sveredyuk) [inline components/courses/config/routes.rb]: only show

File: components/courses/app/helpers/admin/courses/journal_helper.rb [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+    module JournalHelper
+      def journal_mark(student, lecture)
+        ::Courses::Journal::StudentMarkForLecture.call(student, lecture)
+      end
+
+      def journal_presence(student, lecture)
+        ::Courses::Journal::StudentPresenceForLecture.call(student, lecture)
+      end
+
+      def courses_journal_data_color(val, word = '')
```
Comment (sveredyuk) [inline components/courses/app/helpers/admin/courses/journal_helper.rb]: value

File: components/courses/app/helpers/admin/courses/journal_helper.rb [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+        # TODO: move to css
+        if val.nil?
+          '#ffffff'
+        elsif val > 0
+          '#e0ffe7'
+        elsif val == 0
+          if word == 'm'
+            '#fffde0'
+          else
+            '#ffffff'
```
Comment (sveredyuk) [inline components/courses/app/helpers/admin/courses/journal_helper.rb]: WHITE

Comment (sveredyuk) [reply components/courses/app/helpers/admin/courses/journal_helper.rb]:
```ruby

NEGATIVE = '#ffe8e0'
POSITIVE = '#e0ffe7'
WARNING  = '#fffde0'
NEUTRAL  = '#ffffff'

def courses_journal_data_color(val, mark: false)
  return NEUTRAL  unless value
  return NEGATIVE if value < 0
  return POSITIVE if value > 0
  return WARNING  if mark && value == 0

  NEUTRAL
end
```

**Review by sveredyuk — APPROVED** [review]

**Pull request merged**